pkgs/pkgs-lib/tests/formats.nix

```diff
@@ -31,12 +28,61 @@ let
     fi
   '';
 
-  runBuildTests = tests: pkgs.linkFarm "nixpkgs-pkgs-lib-format-tests" (mapAttrsToList (name: value: { inherit name; path = runBuildTest name value; }) (filterAttrs (name: value: value != null) tests));
+  # use this to check for proper serialization
+  shouldPass = format: input: expected: name: {
+    name = "pass-${name}";
+    path = runDiff "test-format-${name}" (format.generate "test-format-${name}" (mergeInput name format input)) expected;
+  };
+
+  # use this function to assert that a type check must fail
+  # note that as per 352e7d330a26 and 352e7d330a26 the type checking of attrsets and lists are not strict
+  # this means that the code below needs to properly merge the module type definition and also evaluate the (lazy) return value
+  shouldFail = format: input: name:
+    let
+      # trigger a deep type check using the module system
+      typeCheck = lib.modules.mergeDefinitions
+        [ "tests" name ]
+        format.type
+        [
+          {
+            file = "format-test-${name}";
+            value = input;
+          }
+        ];
+      # actually use the return value to trigger the evaluation
+      eval = builtins.tryEval (typeCheck.mergedValue == input);
+      # the check failing is what we want, so don't do anything here
+      typeFails = pkgs.runCommand "test-format-${name}" {} "touch $out";
+      # bail with some verbose information in case the type check passes
+      typeSucceeds = pkgs.runCommand "test-format-${name}" {
+          passAsFile = [ "inputText" ];
+          testName = name;
+          # toXML seems to be the only way to stringify input that is generic over *any* data type
+          inputText = builtins.toXML input;
```

Comment:
Not a fan of this hack, it makes the output hard to read, e.g.:

```
The following data was used as input:

<?xml version='1.0' encoding='utf-8'?>
<expr>
  <attrs>
    <attr name="foo">
      <attrs>
        <attr name="function">
          <int value="1" />
        </attr>
      </attrs>
    </attr>
  </attrs>
</expr>
```

I believie using `toJSON` instead is fine here, because the only type it can't handle is functions, which we shouldn't get here because if the type check succeeds it should be a serialisable type, since `pkgs.formats` shouln't generally work for non-serialisable types.

Reply:
I was thinking along the lines of "if anyone ever passed a function by accident it'd be really annoying to debug if the failure was in the `builtins.toJSON` call", but since you're advocating for it I'll go with it in the hopes of nobody ever writing *tests* with accidental functions in there ^^